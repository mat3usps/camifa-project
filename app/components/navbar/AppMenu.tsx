import { Link } from "@remix-run/react";
import classnames from "classnames";
import type { Menu } from "~/utils/appMenus";
import APP_MENUS from "~/utils/appMenus";

interface IAppMenuProps {
  type: "horizontal" | "vertical";
}

const AppMenu = ({ type }: IAppMenuProps) => {
  const isVertical = type === "vertical";
  const isHorizontal = type === "horizontal";

  return (
    <ul
      tabIndex={0}
      className={classnames("menu", {
        "menu-horizontal p-0": isHorizontal,
        "dropdown-content rounded-box menu-compact mt-3 w-52 bg-base-100 p-2 shadow":
          isVertical,
      })}
    >
      {APP_MENUS.map(renderMenu)}
    </ul>
  );

  function renderMenu({ label, subMenus, to }: Menu) {
    if (subMenus) {
      return (
        <li key={label} tabIndex={0}>
          <span className="justify-between">
            {label}
            {renderWithSubMenuIcon()}
          </span>
          {renderSubMenus(subMenus)}
        </li>
      );
    }
    if (to) {
      return (
        <li key={label}>
          <Link tabIndex={0} to={to}>
            {label}
          </Link>
        </li>
      );
    }
    return null;
  }

  function renderSubMenus(subMenus: Menu[]) {
    return (
      <ul className="bg-base-100 p-2 shadow">
        {subMenus.map(({ label, to }) => {
          if (to) {
            return (
              <li key={label}>
                <Link to={to}>{label}</Link>
              </li>
            );
          }
          return null;
        })}
      </ul>
    );
  }

  function renderWithSubMenuIcon() {
    if (isVertical) {
      return (
        <svg
          className="fill-current"
          xmlns="http://www.w3.org/2000/svg"
          width="24"
          height="24"
          viewBox="0 0 24 24"
        >
          <path d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z" />
        </svg>
      );
    }
    return (
      <svg
        className="fill-current"
        xmlns="http://www.w3.org/2000/svg"
        width="20"
        height="20"
        viewBox="0 0 24 24"
      >
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" />
      </svg>
    );
  }
};

export default AppMenu;
