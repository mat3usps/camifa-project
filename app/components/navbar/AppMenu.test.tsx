/* eslint-disable jest/no-conditional-expect */

import { render, screen } from "@testing-library/react";
import type { ComponentProps } from "react";
import { MemoryRouter } from "react-router-dom";
import APP_MENUS from "~/utils/appMenus";
import AppMenu from "./AppMenu";

const { getByRole, getByText, queryByText } = screen;

describe("<AppMenu />", () => {
  it("should all app Menus", () => {
    getRenderer({ type: "vertical" });

    APP_MENUS.forEach(({ label, subMenus, to }) => {
      if (subMenus) {
        const menu = getByText(label);

        expect(menu).toBeInTheDocument();
        subMenus.forEach((subMenu) => {
          if (subMenu.to) {
            expect(
              getByRole("link", { name: subMenu.label })
            ).toBeInTheDocument();
          } else {
            expect(queryByText(subMenu.label)).not.toBeInTheDocument();
          }
        });
      } else if (!to) {
        expect(queryByText(label)).not.toBeInTheDocument();
      } else {
        expect(getByRole("link", { name: label })).toBeInTheDocument();
      }
    });
  });

  it("should match snapshot for vertical menu", () => {
    const { container } = getRenderer({ type: "vertical" });
    expect(container).toMatchSnapshot();
  });

  it("should match snapshot for horizontal menu", () => {
    const { container } = getRenderer({ type: "horizontal" });
    expect(container).toMatchSnapshot();
  });
});

// Helpers
function getRenderer(props: ComponentProps<typeof AppMenu>) {
  return render(
    <MemoryRouter>
      <AppMenu {...props} />
    </MemoryRouter>
  );
}
