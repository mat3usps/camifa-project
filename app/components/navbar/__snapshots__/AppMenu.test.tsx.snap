// Vitest Snapshot v1

exports[`<AppMenu /> > should match snapshot for horizontal menu 1`] = `
<div>
  <ul
    class="menu menu-horizontal p-0"
    tabindex="0"
  >
    <li>
      <a
        href="/app/contas-bancarias"
        tabindex="0"
      >
        Contas bancárias
      </a>
    </li>
  </ul>
</div>
`;

exports[`<AppMenu /> > should match snapshot for vertical menu 1`] = `
<div>
  <ul
    class="menu dropdown-content rounded-box menu-compact mt-3 w-52 bg-base-100 p-2 shadow"
    tabindex="0"
  >
    <li>
      <a
        href="/app/contas-bancarias"
        tabindex="0"
      >
        Contas bancárias
      </a>
    </li>
  </ul>
</div>
`;
